language/move-lang/src/expansion/translate.rs

```diff
@@ -1037,3 +1046,157 @@ fn assign_unpack_fields(
+
+fn unbound_names_spec_block(unbound: &mut BTreeSet<Name>, sp!(_, sb_): &E::SpecBlock) {
+    sb_.members
+        .iter()
+        .for_each(|member| unbound_names_spec_block_member(unbound, member))
+}
+
+fn unbound_names_spec_block_member(unbound: &mut BTreeSet<Name>, sp!(_, m_): &E::SpecBlockMember) {
+    use E::SpecBlockMember_ as M;
+    match m_ {
+        M::Condition { exp, .. } | M::Invariant { exp, .. } => unbound_names_exp(unbound, exp),
+        // No unbound names
+        // And will error in the move prover
+        M::Function { .. } | M::Variable { .. } => (),
+    }
+}
+
+fn unbound_names_exp(unbound: &mut BTreeSet<Name>, sp!(_, e_): &E::Exp) {
+    use E::Exp_ as EE;
+    match e_ {
+        EE::Value(_)
+        | EE::InferredNum(_)
+        | EE::Break
+        | EE::Continue
+        | EE::UnresolvedError
+        | EE::Unit => (),
+        EE::Copy(v) | EE::Move(v) => {
+            unbound.insert(v.0.clone());
+        }
+        EE::Name(n) => {
+            unbound.insert(n.clone());
+        }
+        EE::GlobalCall(_, _, sp!(_, es_)) | EE::Call(_, _, sp!(_, es_)) => {
+            unbound_names_exps(unbound, es_)
+        }
+        EE::Pack(_, _, es) => unbound_names_exps(unbound, es.iter().map(|(_, (_, e))| e)),
+        EE::IfElse(econd, et, ef) => {
+            unbound_names_exp(unbound, ef);
+            unbound_names_exp(unbound, et);
+            unbound_names_exp(unbound, econd)
+        }
+        EE::While(econd, eloop) => {
+            unbound_names_exp(unbound, eloop);
+            unbound_names_exp(unbound, econd)
+        }
+        EE::Loop(eloop) => unbound_names_exp(unbound, eloop),
+
+        EE::Block(seq) => unbound_names_sequence(unbound, seq),
+        EE::Lambda(ls, er) => {
+            unbound_names_exp(unbound, er);
+            // remove anything in `ls`
+            unbound_names_binds(unbound, ls);
+        }
+        EE::Assign(ls, er) => {
+            unbound_names_exp(unbound, er);
+            // remove anything in `ls`
+            unbound_names_assigns(unbound, ls);
+        }
+        EE::Return(e)
+        | EE::Abort(e)
+        | EE::Dereference(e)
+        | EE::UnaryExp(_, e)
+        | EE::Borrow(_, e)
+        | EE::Cast(e, _)
+        | EE::Annotate(e, _) => unbound_names_exp(unbound, e),
+        EE::FieldMutate(ed, er) => {
+            unbound_names_exp(unbound, er);
+            unbound_names_dotted(unbound, ed)
+        }
+        EE::Mutate(el, er) | EE::BinopExp(el, _, er) => {
+            unbound_names_exp(unbound, er);
+            unbound_names_exp(unbound, el)
+        }
+        EE::ExpList(es) => unbound_names_exps(unbound, es),
+        EE::ExpDotted(ed) => unbound_names_dotted(unbound, ed),
+        EE::Index(el, ei) => {
+            unbound_names_exp(unbound, ei);
+            unbound_names_exp(unbound, el)
+        }
+
+        EE::Spec(_, unbound_names) => unbound.extend(unbound_names.iter().cloned()),
```

Comment:
This would be nested SpecBlock, I suppose. Most likely we do not allow this.  @wrwg ?